xarray/core/coordinates.py

```diff
@@ -796,6 +797,126 @@ def _ipython_key_completions_(self):
         ]
 
 
+class DataTreeCoordinates(DatasetCoordinates):
+    """
+    Dictionary like container for coordinates of a DataTree node (variables + indexes).
+    """
+
+    # TODO: This collection can be passed directly to the :py:class:`~xarray.Dataset`
+    # and :py:class:`~xarray.DataArray` constructors via their `coords` argument.
+    # This will add both the coordinates variables and their index.
+
+    # TODO: This only needs to be a separate class from `DatasetCoordinates` because DataTree nodes store their variables differently
+    # internally than how Datasets do, see https://github.com/pydata/xarray/issues/9203.
+
+    # TODO should inherited coordinates be here? It would be very hard to allow updating them...
+    # But actually maybe the ChainMap approach would make this work okay??
```

Comment:
I can't see an easy way to make assignment of coordinates to parent nodes work. I think it would be better just to have `dt.coords` constructed from inherited coords, but not allow setting coordinates further up the tree. I believe that's consistent with the behaviour we currently have for `dt.__setitem__` anyway.

Reply:
Agreed!